src/weight_windows.cpp

```diff
@@ -0,0 +1,314 @@
+    if (value.size() != biasing_energy.size() - 1) {
+      fatal_error("Biasing and biasing energies must have the same number of values.");
+    }
+    biasing.push_back(0);
+    cumulative_biasing.push_back(0);
+    for (double v : value) {
+      biasing.push_back(v);
+      cumulative_biasing.push_back(0);
+    }
+    // normalization
+    double total_probability = std::accumulate(biasing.begin(), biasing.end(), 0.0);
+    for (int i = 1; i < biasing.size(); i++) {
+      biasing[i] = biasing[i]/total_probability;
+      cumulative_biasing[i] = cumulative_biasing[i-1] + biasing[i];
+    }
+  } else {
+    fatal_error("Must provide biasing for each energy group.");
+  }
+}
+  
+//! source weight biasing in energy
+void WeightWindow::weight_biasing(SourceSite& site, uint64_t* seed) 
+{
+  int i = 0;
+  double r = prn(seed);
+  for (i = 0; i < cumulative_biasing.size() - 1; i++) 
+    if (cumulative_biasing[i] <= r && r < cumulative_biasing[i+1])  break;
+  site.E = biasing_energy[i] + ( biasing_energy[i+1] - biasing_energy[i] ) * prn(seed);
+  site.wgt *= origin_probability[i+1] / biasing[i+1];
+}
+  
+//! Get weight windows parameters given particle - essentially
+// given a location tell the particle the right bounds
+ParticleWeightParams WeightWindow::get_params(Particle& p) const
+{	
+  // Particle's position and energy
+  Position pos  = p.r();
+  double E = p.E();
+
+  // get the settings for the weight window
+  ParticleType type = p.type();
+
+  // 
+  WWParams ww_settings;
+  auto it = weight_params.find(p.type());
+  // return if
+
+  // if we dont find the particle in the map return
+  // in the window - no spliting will happen
+  if( it != weight_params.end()) {
+    ww_settings = it->second;
+  } else {
+    // no ww settings found - return in
+    return (ParticleWeightParams){0,1,0.5,1,1e-12};
+  }
+	
+  // indices in weight window vector
+  int indices = mesh_->get_bin(pos);
+
+  // no ww settings found - return in
+  if ( indices < 0 ) return (ParticleWeightParams){0,1,0.5,1,1e-12};
+
+    // get the mesh bin in energy group
+  int energy_bin = lower_bound_index(ww_settings.energy_bounds.begin(), 
+                     ww_settings.energy_bounds.end(), E);
+  // indices points to the correct weight given 
+  // an energy
+  indices += energy_bin*mesh_->n_bins(); 
+
+  ParticleWeightParams params;
+  // set the weight for the current location
+  params.lower_weight = ww_settings.multiplier*
+                         ww_settings.lower_ww[indices];  
+  // set the upper weight bound
+  params.upper_weight = ww_settings.upper_ratio*
+                         params.lower_weight;
+  // set the survival weight
+  params.survival_weight = params.lower_weight*
+                          ww_settings.survival_ratio;
+  // set the max split
+  params.max_split = ww_settings.max_split;
+
+  // set the weight cutoff
+  params.weight_cutoff = ww_settings.weight_cutoff;
```

Comment:
We could probably make this a bit cleaner by just having a constructor `ParticleWeightParams(ww_settings, indices)`